Fix enum value's __int__ returning non-int when underlying type is bool or of char type

Vigilans opened the pull request:
### Introduction

This PR solves the problem of enum value's `__int__` returning a non-int value when its underlying type is bool or of char type.

### Detail

This PR extends the `enum_` class with following type alias:

* `Underlying` type: the actual underlying type of enum (The original `Scalar` type);
* `Scalar` type: the integer type with **length** and **signedness** equivalent to the underlying type.

The enum should always use `Scalar` as the inner integer type when dealing with python side.

So,
* Scalar for underlying type of normal integer would stay the same; 
* Scalar for underlying type of bool or char type would use the int type of same length and signedness, specified through the `equivalent_integer` struct.

### Applicable issues
Fix #1331
Fix #1820
Fix https://github.com/osmcode/pyosmium/issues/114

jagerman commented:
Is `unsigned char` (or, for that matter, `signed char`, which is also distinct from `char` and `unsigned char`) necessary here?  That `is_std_char_type` template is there to capture the types that are used for strings, which is pretty much just `char`.  It shouldn't be catching `unsigned char` and `signed char` in the first place--those should fall through to the arithmetic converter.

Vigilans-Yea commented:
@jagerman 

I see. It corresponds to the compile error which seems to tell me that `unsigned char` is used as an arithmetic type, instead of an element of string, which is listed inside the `is_std_char_type`.
Thank you for your explanation≡ω≡

Vigilans commented:
Sorry for my foolishness of not getting your point and suspended this PR for one year long...

I've updated the PR to keep pace with the master, and fixed it with a more robust mechanism, along with more tests, to solve both issues of char and bool types.

My original post for context reference:

> I followed the suggestion mentioned in the issue and now the bug is eliminated.
> 
> BTW, It should be noted that the definition of `detailed::is_std_char_type` does not check  `unsigned char`:
> ```cpp
> template <typename CharT> using is_std_char_type = any_of<
>     std::is_same<CharT, char>, /* std::string */
>     std::is_same<CharT, char16_t>, /* std::u16string */
>     std::is_same<CharT, char32_t>, /* std::u32string */
>     std::is_same<CharT, wchar_t> /* std::wstring */
> >;
> ```
> And when I tried to add `unsigned char` to the type list, compile errors occurred. Therefore, I alternatively turn to add it to the definition of `enum_<Type>::ScalarInt`:
> ```cpp
> using ScalarInt = detail::conditional_t<
>     detail::is_std_char_type<Scalar>::value || std::is_same<Scalar, unsigned char>::value,
>     detail::conditional_t<std::is_signed<Scalar>::value, int, unsigned>,
>     Scalar>;
> ```

Vigilans commented:
Hi @wjakob @jagerman, now CI has been fixed, would you please take some time to review it? Sorry for suspending this PR for too long...

ax3l commented:
HI @Vigilans - looking through unfixed problems with propose solutions, do you think it would be possible to rebase this pull request for a new round of review? Sorry for the long delay on this!
